codalab/server/authenticated_plugin.py

```diff
@@ -59,16 +94,44 @@ class AuthenticatedPlugin(object):
       - Requests from the CLI.
     """
 
-    api = 2
+    def apply(self, callback, route):
+        def wrapper(*args, **kwargs):
+            self.check_user_authenticated()
+            return callback(*args, **kwargs)
+
+        return wrapper
+
+
+class AuthenticatedProtectedPlugin(AuthenticationPlugin):
+    """
+    Fails the request if the user is not authenticated. It doesn't matter if the instance is protected or not.
+    Then checks if the user has access to the instance if the instance is protected.
+
+    The handling of AJAX requests is the same as AuthenticatedPlugin.
+    """
 
     def apply(self, callback, route):
         def wrapper(*args, **kwargs):
-            if not user_is_authenticated():
-                if request.is_ajax:
-                    abort(httplib.UNAUTHORIZED, 'Not authorized')
-                else:
-                    redirect_with_query('/account/login', {'next': request.url})
+            self.check_user_authenticated()
+            if self.is_protected_mode():
+                self.check_has_access()
+            return callback(*args, **kwargs)
+
+        return wrapper
 
+
+class ProtectedPlugin(AuthenticationPlugin):
+    """
+    Checks if the user is authenticated and has access to the instance only if the instance is in protected mode.
```

Comment:
It was a bit hard to immediately understand what the classes were doing by just reading the comments.

Could we structure all our comments for these classes like this:

```
In non-protected mode, grants anonymous access.

In protected mode, checks if the user is authenticated and has access to the instance.
```